hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala

```diff
@@ -29,4 +30,94 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
+
+    if (nDims == 0) {
+      0L
+    }
+    else {
+      val totalElements = mb.newField[Long]
+      Code(
+        totalElements := 1L,
+        Code.foreach(0 until nDims) { index =>
+          totalElements := totalElements * getShapeAtIdx(index)
+        },
+        totalElements
+      )
+    }
+  }
+
+  def makeDefaultStrides(sourceShapePType: PTuple, sourceShape: Code[Long], mb: MethodBuilder): Code[Long] = {
+    def getShapeAtIdx(index: Int) = Region.loadLong(sourceShapePType.loadField(sourceShape, index))
+
+    val stridesPType = this.representation.fieldType("strides").asInstanceOf[PTuple]
+    val srvb = new StagedRegionValueBuilder(mb, stridesPType)
+    val tupleStartAddress = mb.newField[Long]
+    Code (
+      srvb.start(),
+      tupleStartAddress := srvb.offset,
+      // Fill with 0s, then backfill with actual data
+      Code.foreach(0 until nDims) { index =>
+        Code(srvb.addLong(0L), srvb.advance())
+      },
+      {
+        val runningProduct = mb.newField[Long]
+        Code(
+          runningProduct := elementType.byteSize,
+          Code.foreach((nDims - 1) to 0 by -1) { idx =>
+            val fieldOffset = stridesPType.fieldOffset(tupleStartAddress, idx)
+            Code(
+              Region.storeLong(fieldOffset, runningProduct),
+              runningProduct := runningProduct * getShapeAtIdx(idx)
+            )
+          },
+          srvb.end()
+        )
+      }
+    )
+  }
+
+  def getElementPosition(indices: Seq[ClassFieldRef[Long]], nd: Code[Long], region: Code[Region], mb: MethodBuilder): Code[Long] = {
+    val rep = this.representation
+    val strides = rep.loadField(region, nd, "strides")
+    val dataCode = rep.loadField(region, nd, "data")
+    val dataP = rep.fieldType("data").asInstanceOf[PArray]
+    def getStrideAtIdx(idx: Int): Code[Long] = Region.loadLong(rep.fieldType("strides").asInstanceOf[PTuple].loadField(strides, idx))
+    val bytesAway = mb.newLocal[Long]
+    val data = mb.newLocal[Long]
+    coerce[Long](Code(
+      data := dataCode,
+      bytesAway := 0L,
+      indices.zipWithIndex.foldLeft(Code._empty[Unit]){case (codeSoFar: Code[_], (requestedIndex: ClassFieldRef[Long], strideIndex: Int)) =>
+        Code(
+          codeSoFar,
+          bytesAway := bytesAway + requestedIndex * getStrideAtIdx(strideIndex)
+        )
+      },
+
+      bytesAway + dataP.elementOffset(data, dataP.loadLength(data), 0)
+    ))
+  }
+
+  def construct(flags: Code[Int], offset: Code[Int], shape: Code[Long], strides: Code[Long], data: Code[Long], mb: MethodBuilder): Code[Long] = {
+    val srvb = new StagedRegionValueBuilder(mb, this.representation)
+    val shapeP = this.representation.fieldType("shape").asInstanceOf[PTuple]
+
+    coerce[Long](Code(
+      srvb.start(),
+      srvb.addInt(flags),
+      srvb.advance(),
+      srvb.addInt(offset),
+      srvb.advance(),
+      srvb.addIRIntermediate(this.representation.fieldType("shape").asInstanceOf[PTuple])(shape),
```

Comment:
Do you need the `asInstanceOf[T]` casts here? `addIRIntermediate` should work without them.